external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka/ConsumerStrategy.scala

```diff
@@ -0,0 +1,164 @@
+   */
+  def onStart(currentOffsets: Map[TopicPartition, Long]): Consumer[K, V]
+}
+
+/**
+ * Subscribe to a collection of topics.
+ * @param topics collection of topics to subscribe
+ * @param kafkaParams Kafka
+ * <a href="http://kafka.apache.org/documentation.htmll#newconsumerconfigs">
+ * configuration parameters</a> to be used on driver. The same parameters will be used on executors,
+ * with minor automatic modifications applied.
+ *  Requires "bootstrap.servers" to be set
+ * with Kafka broker(s) specified in host1:port1,host2:port2 form.
+ */
+case class Subscribe[K: ClassTag, V: ClassTag](
+    topics: ju.Collection[java.lang.String],
+    kafkaParams: ju.Map[String, Object]
+  ) extends ConsumerStrategy[K, V] {
+
+  def executorKafkaParams: ju.Map[String, Object] = kafkaParams
+
+  def onStart(currentOffsets: Map[TopicPartition, Long]): Consumer[K, V] = {
+    val consumer = new KafkaConsumer[K, V](kafkaParams)
+    consumer.subscribe(topics)
+    consumer
+  }
+}
+
+object Subscribe {
+  def create[K, V](
+      keyClass: Class[K],
+      valueClass: Class[V],
+      topics: ju.Collection[java.lang.String],
+      kafkaParams: ju.Map[String, Object]
+  ): Subscribe[K, V] = {
+    implicit val keyCmt: ClassTag[K] = ClassTag(keyClass)
+    implicit val valueCmt: ClassTag[V] = ClassTag(valueClass)
+    Subscribe[K, V](topics, kafkaParams)
+  }
+}
+
+/**
+ * Assign a fixed collection of TopicPartitions
+ * @param topicPartitions collection of TopicPartitions to subscribe
+ * @param kafkaParams Kafka
+ * <a href="http://kafka.apache.org/documentation.htmll#newconsumerconfigs">
+ * configuration parameters</a> to be used on driver. The same parameters will be used on executors,
+ * with minor automatic modifications applied.
+ *  Requires "bootstrap.servers" to be set
+ * with Kafka broker(s) specified in host1:port1,host2:port2 form.
+ */
+case class Assign[K: ClassTag, V: ClassTag](
+    topicPartitions: ju.Collection[TopicPartition],
+    kafkaParams: ju.Map[String, Object]
+  ) extends ConsumerStrategy[K, V] {
+
+  def executorKafkaParams: ju.Map[String, Object] = kafkaParams
+
+  def onStart(currentOffsets: Map[TopicPartition, Long]): Consumer[K, V] = {
+    val consumer = new KafkaConsumer[K, V](kafkaParams)
+    consumer.assign(topicPartitions)
+    consumer
+  }
+}
+
+object Assign {
+  def create[K, V](
+      keyClass: Class[K],
+      valueClass: Class[V],
+      topicPartitions: ju.Collection[TopicPartition],
+      kafkaParams: ju.Map[String, Object]
+  ): Assign[K, V] = {
+    implicit val keyCmt: ClassTag[K] = ClassTag(keyClass)
+    implicit val valueCmt: ClassTag[V] = ClassTag(valueClass)
+    Assign[K, V](topicPartitions, kafkaParams)
+  }
+}
+
+
+/**
+ * Set offsets on initial startup only, after another strategy has configured consumer
+ * @param offsets: offsets to begin at
+ * @param init: ConsumerStrategy responsible for instantiation and initial config
+ */
+case class FromOffsets[K: ClassTag, V: ClassTag](
```

Comment:
Yes you can specify starting offsets for any subscription or assignment, because its just a seek on particular topic partitions.
There's a way to subscribe based on a regex, so there's at least one more obvious SubscribePattern strategy subclass.
I can do SubscribeFromOffsets, AssignFromOffsets, SubscribePatternFromOffsets, but this gets to be a combinatorial issue pretty quickly.